blog/2022-04-25-jwt-vs-api-key-authentication/index.md

```diff
@@ -0,0 +1,109 @@
+}
+```
+
+However, most API Key authentication systems offer more advanced management of keys. For example, an API Gateway could offer a self-serve portal where end-users issue tokens based on permissions and configuration that the API developer has set. Tokens can be issued with various permissions and with custom expirations times.
+
+A typical API Key authentication system will validate each key as it comes in with a request. If the key is valid, then data is returned with that key. This data could be identity information, permissions, and even details about their allowed use of the API, such as rate limits or IP restrictions.
+
+```
+function myApiHandler(request) {
+  const apiKey = request.headers.get("API-Key");
+  const apiKeyInfo = apiKeyService.validate(apiKey);
+
+  // Check various properties of the api key info
+  if (clientInfo?.accountId) {
+    // ...
+  }
+}
+```
+
+### Layered Permissions
+
+One of the most powerful capabilities of API Key authentication is the ability to build "layered" permission capabilities into your API. For example, you, as the API owner, could authorize a customer to create API Keys with permissions like `read:products` and `write:products`. The customer could go into their self-serve portal and issue a token with both permissions, but they could also issue a token with a sub-set of permissions. The benefit of this layered approach is that each token has only the permissions needed for its specific function.
+
+![Token Scopes](./token_scopes.gif)
+
+### Considerations of API Key Auth
+
+The main difference between API Key auth and JWT token auth is that the JWT Token is self-contained - the information asserted by the token is in the token. Whereas with an API Key the asserted information is stored in an external system. The externalization of assertion data makes API Keys more flexible for certain scenarios.
+
+- Simpler for end-users to use and understand.
+- Individual keys can be revoked.
+- Permissions and expiration times of keys can be changed even after they are issued.
+- Permissions can be layered/customized at different levels
+- Because the key doesn't contain any information, the associated data for each key can effectively be limitless. For example, an API Key Authentication system could also assert that a particular token is allowed to access a particular account.
+- API Keys can be issued without expirations and revoked only when needed (i.e., a customer cancels their account).
+
+## Summary
+
+Both JWT authentication and API Key authentication are helpful tools when building a secure API. Each one has benefits and drawbacks. API Key authentication tends to be extremely simple for end-users to understand and get started using your API. Still, it is not a standardized protocol and can be more challenging to implement. JWT tends to be more difficult for end-users, but because it is standards-based and so common, it is likely already supported by the framework in which you build your APIs.
```

Comment:
I feel like we should come out MUCH stronger that API key auth is recommended for external API sharing - and reference stripe, twilio as examples. I would talk about importance of allowing developers to roll keys themselves, that's what makes them secure.